File: frontend/src/types/Product.ts
import {StockUnit} from './StockUnit';
import {Category} from './Category';
import {Brand} from './Brand';

export interface Product {
  id: number;
  name: string;
  description: string | null;
  stock_unit: StockUnit;
  weight_product: boolean;
  stock_value: number;
  stock_quantity: number;
  price: number;
  image: string | null;
  category_id: number;
  brand_id: number;
  category?: Category;
  brand?: Brand;
}

File: backend/src/core/schemas/product.py
from typing import Annotated

from annotated_types import MaxLen, Gt, Ge
from fastapi import UploadFile
from pydantic import BaseModel, ConfigDict

from core.schemas.brand import BrandRead
from core.schemas.category import CategoryRead
from utils.enums import StockUnitEnum


class ProductBase(BaseModel):
    name: Annotated[str, MaxLen(255)]
    description: Annotated[str | None, MaxLen(1000)] = None
    stock_unit: StockUnitEnum
    weight_product: bool = False
    stock_value: Annotated[float, Gt(0)]
    stock_quantity: Annotated[float, Ge(0)]
    price: Annotated[int, Ge(0)]


class ProductCreateUpdate(ProductBase):
    category_id: int
    brand_id: int


class ProductListRead(ProductBase):
    model_config = ConfigDict(
        from_attributes=True,
    )

    id: int
    image_url: str | None
    category_id: int
    brand_id: int


class ProductDetailRead(ProductBase):
    model_config = ConfigDict(
        from_attributes=True,
    )

    id: int
    image_url: str | None
    category: CategoryRead
    brand: BrandRead


class ProductResponse(ProductBase):
    model_config = ConfigDict(
        from_attributes=True,
    )

    id: int
    image: str | None
    category_id: int
    brand_id: int


class ProductImageUpload(BaseModel):
    image: UploadFile
